File: server/utils/envVarPhase.js
const BUILD_ARG_KEY_PREFIX = /^(VITE_|NEXT_PUBLIC_|REACT_APP_)/;
const SENSITIVE_KEY_PATTERN = /(SECRET|TOKEN|PASSWORD|PRIVATE|API_KEY)/i;

const ENV_PHASES = ["runtime", "build"];

/**
 * Infer build vs runtime from variable name (Vite, Next, CRA conventions).
 */
function inferEnvPhase(key) {
  const name = String(key || "").trim();
  if (!name) return "runtime";
  if (BUILD_ARG_KEY_PREFIX.test(name)) return "build";
  return "runtime";
}

function normalizeEnvPhase(phase) {
  const p = String(phase || "").trim().toLowerCase();
  return ENV_PHASES.includes(p) ? p : null;
}

/**
 * Secrets and sensitive names must never be Docker build-args (image layer history).
 */
function isSensitiveBuildKey(key) {
  const name = String(key || "").trim();
  if (!name) return true;
  return SENSITIVE_KEY_PATTERN.test(name);
}

/**
 * Resolved phase for a stored env row (explicit phase wins, then inference, then guards).
 */
function resolveEnvPhase(row = {}) {
  const key = String(row.key || "").trim();
  if (!key) return "runtime";

  if (row.isSecret === true || isSensitiveBuildKey(key)) {
    return "runtime";
  }

  const explicit = normalizeEnvPhase(row.phase);
  if (explicit) return explicit;

  return inferEnvPhase(key);
}

/**
 * Split decrypted env rows into Docker build-args vs container runtime env maps.
 */
function splitEnvVarsForDeploy(rows = []) {
  const buildArgs = {};
  const runtimeEnv = {};

  for (const row of rows) {
    const key = String(row?.key || "").trim();
    if (!key) continue;

    const value = row.value ?? "";
    if (resolveEnvPhase(row) === "build") {
      buildArgs[key] = value;
    } else {
      runtimeEnv[key] = value;
    }
  }

  return { buildArgs, runtimeEnv };
}

/**
 * Normalize phase on ingest (UI / API).
 */
function normalizeEnvRowPhase(row = {}) {
  const key = String(row.key || "").trim();
  const explicit = normalizeEnvPhase(row.phase);
  let phase = explicit || inferEnvPhase(key);

  if (row.isSecret === true || isSensitiveBuildKey(key)) {
    phase = "runtime";
  }

  return phase;
}

module.exports = {
  ENV_PHASES,
  BUILD_ARG_KEY_PREFIX,
  inferEnvPhase,
  normalizeEnvPhase,
  resolveEnvPhase,
  isSensitiveBuildKey,
  splitEnvVarsForDeploy,
  normalizeEnvRowPhase,
};

File: agent/app/utils/env_var_phase.py
"""Build-time vs runtime environment variable classification."""

import json
import re
from typing import Any, Dict, Optional, Tuple

_BUILD_ARG_PREFIX = re.compile(r"^(VITE_|NEXT_PUBLIC_|REACT_APP_)")
_SENSITIVE_KEY = re.compile(r"(SECRET|TOKEN|PASSWORD|PRIVATE|API_KEY)", re.IGNORECASE)
_VALID_PHASES = frozenset({"runtime", "build"})


def infer_env_phase(key: str) -> str:
    name = str(key or "").strip()
    if not name:
        return "runtime"
    if _BUILD_ARG_PREFIX.match(name):
        return "build"
    return "runtime"


def is_sensitive_build_key(key: str) -> bool:
    name = str(key or "").strip()
    if not name:
        return True
    return bool(_SENSITIVE_KEY.search(name))


def resolve_env_phase(
    key: str,
    *,
    phase: Optional[str] = None,
    is_secret: bool = False,
) -> str:
    """``is_secret`` is ignored for phase (values are always stored encrypted)."""
    del is_secret
    name = str(key or "").strip()
    if not name:
        return "runtime"
    if is_sensitive_build_key(name):
        return "runtime"
    explicit = str(phase or "").strip().lower()
    if explicit == "build":
        return "build"
    inferred = infer_env_phase(name)
    if explicit == "runtime":
        if inferred == "build":
            return "build"
        return "runtime"
    return inferred


def split_env_vars(
    raw: Optional[Dict[str, Any]],
    phases: Optional[Dict[str, str]] = None,
) -> Tuple[Dict[str, str], Dict[str, str]]:
    """
    Partition env vars into (build_args, runtime_env).
    ``phases`` maps key -> "build"|"runtime" when provided by the platform.
    """
    if not raw:
        return {}, {}

    phase_map = phases or {}
    build_args: Dict[str, str] = {}
    runtime_env: Dict[str, str] = {}

    for key, val in raw.items():
        if key is None:
            continue
        ks = str(key).strip()
        if not ks:
            continue
        if val is None:
            vs = ""
        elif isinstance(val, (dict, list)):
            vs = json.dumps(val, separators=(",", ":"))
        else:
            vs = str(val)

        resolved = resolve_env_phase(ks, phase=phase_map.get(ks))
        if resolved == "build":
            build_args[ks] = vs
        else:
            runtime_env[ks] = vs

    return build_args, runtime_env
